### ISSUE TITLE
Qwen3 .with_reasoning(): `thinking` field stays empty when `</think>` is not reached

### ISSUE THREAD

**Issue by shsym** [pie-project/pie#328]
## Qwen3 `.with_reasoning()`: `thinking` field stays empty when `</think>` isn't reached

- Branch: `features/deva-integration`
- Inferlet: `inferlets/text-completion`
- Model: `Qwen/Qwen3-0.6B`, vllm driver, FlashInfer attention

### Observed

When generation is cut off (`max_tokens` reached) before the model emits `</think>`, the inferlet's `thinking` field is empty even though the model clearly produced reasoning content — the reasoning text instead lands in `text`, prefixed with the literal `<think>` characters.

Concrete output for *"What is 2 + 2? Answer in one word."* with `--max_tokens 32 --temperature 0.6 --top_p 0.95`:

```json
{
  "thinking": "",
  "text": "<think>\nOkay, the user is asking what 2 plus 2 is, and they want the answer in one word. Let me think. 2 plus"
}
```

The same prompt run through `text-completion-spec` (the lower-level `ctx.generate(sampler).next()` API, no decoder layer) produces the identical raw byte stream starting with `<think>\n`, so the model is emitting reasoning tokens correctly — the gap is on the SDK-side `ThinkingDecoder`/`EventStream::with_reasoning` path.

### Why this matters

For an interactive UI that wants to stream reasoning incrementally, or for any inferlet that caps `max_tokens` to bound latency, this means **partial reasoning is never observable** through `Event::Thinking`. The user sees an empty `thinking` and a `text` that includes the `<think>` tag literally.

### Where the matching happens

[`runtime/src/model/instruct/qwen3.rs#L178-L179`](https://github.com/pie-project/pie/blob/features/deva-integration/runtime/src/model/instruct/qwen3.rs#L178-L179):

```rust
let think_prefix = encode("<think>");
let think_suffix = encode("</think>");
```

These ids are matched against the live token stream in [`runtime/src/model/instruct/decoders.rs#L80-L118`](https://github.com/pie-project/pie/blob/features/deva-integration/runtime/src/model/instruct/decoders.rs#L80-L118) (`ThinkingDecoder::feed`).

I haven't verified empirically whether the bug also manifests when `</think>` IS reached — only that the cut-off case is broken. Two possibilities:

1. **`<think>` matching fails entirely** (e.g. special-token vs subword cardinality mismatch in `encode`), so the decoder never enters `inside=true` state and reasoning content silently leaks to `Event::Text` regardless of whether `</think>` arrives.
2. **`<think>` matching works but `Delta` events emit empty strings** (e.g. tokenizer.decode of single tokens often yields `""` due to multi-token-character boundaries), so `thinking` only fills via `Complete(accumulated)` when `</think>` is reached.

Either case is consistent with what we observed; a longer smoke that lets the model reach `</think>` would disambiguate.

### Reproducer

```bash
pie config init
# In ~/.cache/pie/config.toml under [model.default.driver.vllm]:
#   attention_backend = "FLASHINFER"
#   gpu_memory_utilization = 0.8
#   max_num_batched_tokens = 10240
#   max_num_seqs = 512
pie model download Qwen/Qwen3-0.6B
pie run --path inferlets/text-completion/target/wasm32-wasip2/release/text_completion.wasm \
    --manifest inferlets/text-completion/Pie.toml -- \
    --prompt "What is 2 + 2? Answer in one word." \
    --max_tokens 32
```

### Suggested first step

A unit test alongside [`runtime/src/model/instruct/qwen3.rs::tests`](https://github.com/pie-project/pie/blob/features/deva-integration/runtime/src/model/instruct/qwen3.rs) that constructs `QwenInstruct` with the real Qwen3 tokenizer and asserts:

- `tokenizer.decode(&think_prefix_ids, false) == "<think>"`
- feeding `think_prefix_ids` through `ThinkingDecoder::feed` yields `ReasoningEvent::Start`
- feeding subsequent reasoning tokens yields `ReasoningEvent::Delta(non_empty_string)`

If any of those fail, that pinpoints which of the two failure modes above is real, and the fix follows. Happy to put up a PR once we agree on the right shape.

---

Context: end-to-end token generation works via `text-completion-spec`, so this is the last gap on the canonical reasoning-aware inferlet for Qwen3.

**Comment by shsym** [comment]
## Update: encode-mismatch hypothesis FALSIFIED — root cause needs runtime verification

Ran a focused probe against pie's runtime tokenizer (`runtime/src/model/tokenizer.rs`, branch `features/deva-integration`) loaded directly from `Qwen/Qwen3-0.6B`'s `tokenizer.json`. The previously suspected encode-mismatch is **NOT** the root cause:

```
encode("<think>"):   pie=[151667]  hf=[151667]   match=true
encode("</think>"):  pie=[151668]  hf=[151668]   match=true
encode("<|im_start|>"): pie=[151644]  hf=[151644]   match=true
decode([151667]) = "<think>"
decode([151668]) = "</think>"
```

So `think_prefix_ids = [151667]` at `qwen3.rs:178`, exactly aligned with what HF tokenizers emits and what the model's vocab expects. `ThinkingDecoder::feed(&[151667])` will deterministically transition `inside → true` and return `ReasoningEvent::Start` (verified by the existing decoder unit tests in `decoders.rs::tests::thinking_full_cycle`).

### Pure-logic chain from the observation

The observed output is `{"thinking": "", "text": "<think>\nOkay, ..."}`. Two facts:
1. `thinking == ""` ⟹ `Event::Thinking(non_empty)` never fired
2. `<think>\n` reached `text` ⟹ `Event::Text` fired during those batches ⟹ at the chat-fallback in [decoder.rs#L154-L158](https://github.com/pie-project/pie/blob/features/deva-integration/sdk/rust/inferlet/src/context/decoder.rs#L154-L158), `state` was **not** `Reasoning`

Therefore: `ThinkingDecoder::feed` never returned `Start`, i.e. its `start_ids` were never matched against the live token stream. Since `start_ids = [151667]` is correct (probe above), the only consistent explanation is that **the live token stream from the model does not contain id 151667 at the start of reasoning**.

### Most plausible remaining cause

The model is emitting **subwords** for `<think>` instead of the special token id 151667. e.g. `[27, 26865, 29]` (`<` + `think` + `>`) which `tokenizer.decode(...)` collapses back to the literal string `<think>`, fooling the prefix-of-text observation while leaving `ThinkingDecoder`'s id-level matcher blind.

This would also explain why `text-completion-spec` (no decoder layer, just `tokenizer.decode(all_ids)`) shows the identical leading bytes — both code paths would render the same way regardless of whether the model picked the special-token id or a subword sequence.

### What this means for the fix

The fix is **not** in the SDK decoder pipeline or in `qwen3.rs` `encode("<think>")`. It needs to land further upstream — most likely in how the prompt is presented to the model or how the sampler/logits-processor treats the special-token id. Possible directions:

1. Verify what id the model actually emits at the first decode step. Add a temporary instrumented inferlet (or logging in the runtime's vllm driver) that prints `Δ ids` per batch alongside the SDK events for the canonical `Qwen/Qwen3-0.6B` + `--max_tokens 32` repro.
2. If the model is emitting subwords, check whether the prompt rendering (`<|im_start|>assistant\n` cue) is being tokenized in a way that doesn't put the model in its training-time state for emitting 151667.
3. Cross-check that vLLM's tokenizer (which encodes the prompt) and pie's runtime tokenizer (which decodes ids) are consistent on the boundary tokens around `<|im_start|>assistant\n`.

I'm pausing the PR-shaped fix until step (1) confirms which id the model actually emits — happy to come back with a concrete patch once that's pinned. The companion unit-test idea from the original "Suggested first step" is still good and I'll include it (with a real-tokenizer fixture) regardless of where the fix lands.

---

Probe code (standalone): used a copy of `runtime/src/model/tokenizer*.rs` against the cached `Qwen/Qwen3-0.6B/tokenizer.json` snapshot, using `tokenizers = "0.21"` for the HF reference. Reproducible with the existing `runtime/tests/tokenizer_compat.rs` infrastructure (auto-discovers HF cache). Ran on `features/deva-integration` @ 82ebccbb.

**Comment by shsym** [comment]
## Root cause found and PR up: #329

After the encode-mismatch hypothesis was [falsified](https://github.com/pie-project/pie/issues/328#issuecomment-4354989095), I dug further and pinned the actual bug.

### What's actually broken

`runtime/src/model/instruct::create()` only matches pie's lowercase internal arch names (`"qwen3"`, `"qwen2"`, …), but `pie_driver_vllm/loader.py:212-220` and `pie_driver_sgl/loader.py:209` forward the HuggingFace `architectures[0]` string verbatim — `"Qwen3ForCausalLM"` for Qwen3-0.6B. That falls through to the `_` arm:

```rust
_ => Arc::new(QwenInstruct::new(tokenizer, ChatMLConfig {
    has_thinking: false,
    has_tools: false,
    ...
})),
```

`has_thinking: false` makes `reasoning_decoder()` return `NoopReasoningDecoder`, whose `feed` always returns `Delta(empty)`. In the SDK Decoder pipeline (`sdk/rust/inferlet/src/context/decoder.rs:107-162`) `Delta(empty)` with `state == Normal` falls through to the chat decoder, which decodes id `151667` to the literal string `"<think>"` and emits it as `Event::Text`. That's exactly the observed `{"thinking": "", "text": "<think>\n…"}`.

### Verification chain

1. **Tokenizer**: pie's runtime `encode("<think>") == [151667]` and `decode([151667]) == "<think>"` — already proven [in the previous comment](https://github.com/pie-project/pie/issues/328#issuecomment-4354989095).
2. **Model output**: ran `Qwen/Qwen3-0.6B` locally on Apple MPS via `transformers.AutoModelForCausalLM` with the exact same params (T=0.6, top_p=0.95, max=32). The model's *first* generated id is `151667`, and `151668` (`</think>`) does not appear in 32 tokens. So the model emits the special token, not subwords. This rules out the alternative "model emits `<` + `think` + `>`" hypothesis.
3. **Decoder pipeline**: simulated `ThinkingDecoder` + `GenericChatDecoder` against the actual id sequence. With **per-token batches** the simulation produces the *correct* output (`thinking` populated, `text` empty). So the matcher and chat decoder logic are fine.
4. **Driver hand-off**: `pie_driver/loader.py:65-83` translates `hf_model_type` → pie internal arch via `HF_TO_PIE_ARCH` (so the legacy hand-rolled path is fine). `pie_driver_vllm/loader.py:220` does `arch_type = arches[0]` from HF — that's the divergence. Verified `Qwen/Qwen3-0.6B/config.json` has `"model_type": "qwen3"` AND `"architectures": ["Qwen3ForCausalLM"]`. Combine with the `match arch_name` arms in `runtime/src/model/instruct.rs:106-128` and the `_` fallback fires.

### Fix

PR: https://github.com/pie-project/pie/pull/329 — adds the unambiguous HF CamelCase aliases (`Qwen3ForCausalLM`, `Qwen2ForCausalLM`, `Qwen2_5ForCausalLM`, `DeepseekV3ForCausalLM`, `GptOssForCausalLM`, `Gemma2/3/3TextForCausalLM`, `Mistral3ForCausalLM`, `Olmo3ForCausalLM`) into the existing match. `LlamaForCausalLM` is intentionally NOT aliased — HF reuses it for llama2/llama3/llama3.1+, so the Python driver must hand off the disambiguated pie internal name there.

Includes two unit tests (synthetic-vocab tokenizer, no GPU needed) that pin the regression: `qwen3_internal_name_enables_thinking` and `qwen3_hf_arch_name_enables_thinking`.

### Broader implications worth noting

- The `_` fallback returns a `QwenInstruct` (ChatML), so right now every *unrecognised* arch silently gets the wrong chat template. Llama / Gemma loaded through vllm or sgl would currently mis-render prompts. The PR fixes Qwen3 specifically, but a follow-up should either error on unknown arches or align `pie_driver_vllm` / `pie_driver_sgl` with `pie_driver`'s `hf_model_type` + `HF_TO_PIE_ARCH` convention.
- `pie/src/pie/capabilities.py:62-65` documents `arch_name` as "first architecture string from the HF config" (CamelCase), but `pie_driver` actually sends the lowercase internal name — the two drivers disagree. Resolving that contract is what would make this Rust-side aliasing eventually unnecessary.

I haven't been able to e2e-verify the fix on hardware yet (no GPU locally; the canonical `deva-trial-fix3` image needs a rebuild against this branch). Happy to wait on a maintainer with GPU or to fold in a runtime-side e2e once the image is rebuilt.

**Comment by shsym** [comment]
E2E verified ✅ — see https://github.com/pie-project/pie/pull/329#issuecomment-4253.

Built #329 onto features/vllm-opt, ran canonical reproducer on ECL (Qwen/Qwen3-0.6B, max_tokens=32, T=0.6, top_p=0.95, FlashInfer):

- **Pre-fix**: `{"thinking":"","text":"<think>\\nOkay, ..."}` (the bug)
- **Post-fix**: `{"thinking":"\\nOkay, ...","text":""}` (correct)

**Comment by shsym** [comment]
Repointed: closed #329 (Rust-side aliases) in favor of #330, which puts the HF→pie name resolution where it belongs (Python-side) and keeps Rust dispatching on the pie internal name only.

E2E verified on ECL — same pre/post output as #329, but the contract is fixed at the right layer this time. Details and rejected-alternatives note in https://github.com/pie-project/pie/pull/330.
